### ISSUE TITLE
Problems upgrading to TFX 0.27.0

### ISSUE THREAD

**Issue by rafaascensao** [kubeflow/pipelines#5137]
### What steps did you take:
Installed Kubeflow Pipelines on GCP via kustomize manifests.
Tried to run the Taxi TFX Demo.

### What happened:
On the first step, I got the error "No module named 'tfx.dsl.components'"

### What did you expect to happen:
To successfully run the TFX Taxi Demo.

### Environment:
How did you deploy Kubeflow Pipelines (KFP)?
Via the kustomize manifests in GCP.

KFP version: 1.4.0-rc.1

/kind bug
/area backend

**Comment by ConverJens** [comment]
@rafaascensao Which version of TFX are you using? the tfx.dsl package was introduced in TFX 0.25.0 I think, and if you are using an older version this error would pop up. Try running updating TFX to latest by running `python3 -m pip install -U tfx`. Anyway, this is a package import issue and most likely has nothing to do with KFP.

**Comment by rafaascensao** [comment]
@ConverJens I am talking about the sample pipeline of TFX Taxi that comes with the deployment (https://github.com/kubeflow/pipelines/blob/master/samples/core/parameterized_tfx_oss/parameterized_tfx_oss.py). If I see the code, it uses a TFX image with the version 0.22.0.

**Comment by ConverJens** [comment]
@rafaascensao Sorry! In that case, the image needs to be updated to tfx>=0.25.0. Ping @Bobgy, do you know who maintains this?

**Comment by Bobgy** [comment]
/assign @numerology @chensun @neuromage 

This seems a P0.

**Comment by NikeNano** [comment]
Dont know why the linking dont work, but PR for fix: https://github.com/kubeflow/pipelines/pull/5165

**Comment by numerology** [comment]
Thanks @NikeNano !

@Bobgy IIRC our integration test should cover this sample, do we know why this is not captured?

**Comment by Bobgy** [comment]
I have the same concern, @Ark-kun do you know why?

/reopen
Let's make sure to also address why it slipped integration tests.

**Comment by Bobgy** [comment]
I tried to upgrade some other TFX related deps, however, I'm getting the following errors in various places:
* when clicking visualizations on `examplevalidator`, I'm seeing error: `NotFoundError: The specified path gs://gongyuan-test/tfx_taxi_simple/6f9ad66d-2897-4e11-8c2c-e3c614920f35/ExampleValidator/anomalies/32/anomalies.pbtxt was not found.`

**Comment by Bobgy** [comment]
Another problem:
KFP depends on absl-py: https://github.com/kubeflow/pipelines/blob/a394f8cdf34de71f305ca1a49220a859c81ea502/sdk/python/requirements.in#L26

conflicts with

tfx==0.27.0

**Comment by Bobgy** [comment]
The third problem I found, https://github.com/kubeflow/pipelines/tree/master/samples/core/tfx-oss sample is still mentioning very old version of TFX. Can someone verify and fix?

**Comment by chensun** [comment]
> Another problem:
> KFP depends on absl-py:
> 
> https://github.com/kubeflow/pipelines/blob/a394f8cdf34de71f305ca1a49220a859c81ea502/sdk/python/requirements.in#L26
> 
> conflicts with
> 
> tfx==0.27.0
> 
> ```
> ERROR: Cannot install absl-py<1 and >=0.11.0 and tfx==0.27.0 because these package versions have conflicting dependencies.
> 
> The conflict is caused by:
>     The user requested absl-py<1 and >=0.11.0
>     tfx 0.27.0 depends on absl-py<0.11 and >=0.9
> ```
> 
> EDIT: I'm going to use absl that works with TFX, let's see whether it passes presubmit & postsubmit tests.
> #5179

This one actually conflicts with KFP setup.py
https://github.com/kubeflow/pipelines/blob/cd55a55c6229e9e734a35c166c002728b8fa4a72/sdk/python/setup.py#L23
It's an easy fix that we should do anyway, but I want to understand where `requirements.txt` is used in this case? A KFP SDK user doesn't need to install from `requirements.txt`.

**Comment by Ark-kun** [comment]
>I want to understand where requirements.txt is used in this case? A KFP SDK user doesn't need to install from requirements.txt.

It's used in automated tests and documentation generation. Some users might be using it as well. `requirements.txt` provides a known good dependency snapshot while `requirements.in` or `setup.py` specify loose top layer dependencies.
See the official documentation: https://packaging.python.org/discussions/install-requires-vs-requirements/#requirements-files

In the future we should make `setup.py` use `requirements.in` instead of duplicating the list.

**Comment by Ark-kun** [comment]
>KFP depends on absl-py:

It looks like a new dependency. Do we need it much?

**Comment by Ark-kun** [comment]
> * when clicking visualizations on `examplevalidator`, I'm seeing error: `NotFoundError: The specified path gs://gongyuan-test/tfx_taxi_simple/6f9ad66d-2897-4e11-8c2c-e3c614920f35/ExampleValidator/anomalies/32/anomalies.pbtxt was not found.`

Does this file exist? What does the validator component produce?

**Comment by chensun** [comment]
> > I want to understand where requirements.txt is used in this case? A KFP SDK user doesn't need to install from requirements.txt.
> 
> It's used in automated tests and documentation generation. Some users might be using it as well. `requirements.txt` provides a known good dependency snapshot while `requirements.in` or `setup.py` specify loose top layer dependencies.
> See the official documentation: https://packaging.python.org/discussions/install-requires-vs-requirements/#requirements-files
> 

My understanding is that `setup.py` is for end users, while `requirements.txt` is for developers. So my question was more like do we need to install `requirements.txt` during tests? I could image we need `requirements-test.txt` for test library dependencies. But other than that, shouldn't we just rely on `setup.py` for runtime dependencies? This will test against what end users will get rather than a stable known-good environment.

> In the future we should make `setup.py` use `requirements.in` instead of duplicating the list. 

I thought we're deprecating `requirements.in` with the recently introduced bot that updates `requirements.txt` automatically?

**Comment by chensun** [comment]
> > KFP depends on absl-py:
> 
> It looks like a new dependency. Do we need it much?

It was recently introduced, but seems likely only for logging purpose. So I think it's avoidable at least for now.

https://github.com/kubeflow/pipelines/blob/665f3ce8ba0ac0cada2b4a1bd6c5bf8414d99bcb/sdk/python/kfp/dsl/artifact.py#L17

https://github.com/kubeflow/pipelines/blob/665f3ce8ba0ac0cada2b4a1bd6c5bf8414d99bcb/sdk/python/kfp/containers/entrypoint.py#L17

**Comment by Ark-kun** [comment]
>My understanding is that setup.py is for end users, while requirements.txt is for developers.

I'm not sure there is a hard distinction like this. As a snapshot of tested dependencies, `requirements.txt` can still be useful for the users. For example, TFX setup if often broken due to self-conflicting dependencies. However a `requirements.txt` can be installed.
Same with the KFP and tests: with `requirements.txt` there is a tested configuration. Unlike setup.py it's not affected by the environment and already installed packages. Before we started using it, we had issues when CI was flaky or installed different versions of packages at random. Especially when testing on different versions of python. Now we have the full package snapshot that it tested and a way to update that snapshot.

>This will test against what end users get

I'm not sure about this. Without requirements.txt our test environment can be very different from the user environments. The user can have very old packages, but our tests will have the latest ones. `requirements.txt` fixes that by letting the users install the same packages as us.

>I thought we're deprecating requirements.in with the recently introduced bot that updates requirements.txt automatically?

Where does the bot get the initial requirements? Many guides advice putting the dependecies outside the setup.py file.

**Comment by Bobgy** [comment]
Thanks, then absl problem is solved, I verified 0.10.0 is compatible with both.

Let me provide more information about the visualization one

**Comment by Bobgy** [comment]
Another new issue is that postsubmit for iris pipeline is failing after upgrade, we need to figure out if the pipeline file needs any update

**Comment by Ark-kun** [comment]
>Let me provide more information about the visualization one

Thank you.

Visualization can be a tough one. It bothers me that our visualizations are version-dependent, but are baked into the backend. We should try to move to the "visualizations as components" vision, so that the visualizations can be versioned independently.

**Comment by chensun** [comment]
> > My understanding is that setup.py is for end users, while requirements.txt is for developers.
> 
> I'm not sure there is a hard distinction like this. 

Right, I didn't mean there's a hard distinction between the two, but just my personal experience/observation  -- if I'm a KFP user, I only care about `pip install kfp`, and probably never bother to go to this GitHub repo and do an additional install step using the `requirements.txt`. 
My point is that what's specified in `setup.py` is the single source of truth for an end user (as I don't think most of them would come and grab the requirements.txt from our GitHub repo). So the tests should try to simulate that as close as possible. Having a safe snapshot reduces or even eliminates test flakiness, but we also lose the opportunity to discover possible dependency issues an end user may hit.

> > I thought we're deprecating requirements.in with the recently introduced bot that updates requirements.txt automatically?
>
> Where does the bot get the initial requirements? Many guides advice putting the dependecies outside the setup.py file.

According to [this](https://github.com/kubeflow/pipelines/pull/5056#issuecomment-770346410), it looks at the existing `requirements.txt` and keep moving forward to the next available minor version.

**Comment by Bobgy** [comment]
@chensun @Ark-kun two PRs pending review: #5186 #5187

**Comment by Bobgy** [comment]
Found a new problem, KFP cache webhook starts to cache tfx pipelines.

**Comment by Bobgy** [comment]
I'm renaming the issue, because we are using it as a collector for all tfx issues.

**Comment by Bobgy** [comment]
All immediate blockers for release have been resolved. I'm going to create separate issues that we can tackle later.
